### PR TITLE
change payment structure of momo-fdi and add airtel payment api

### PR DESCRIPTION
**Pull request by iradukunda1**

### CODE REVIEW

**Comment by rugwirobaker** [comment]
Why did you create new interfaces? The reason we have interfaces in payments is so that we can hide the implementation details of each driver and you the same types for everything.

**Review by rugwirobaker — CHANGES_REQUESTED** [review]
This is a good start but we need to remove the driver(s) package and simplify the `TokenSource` code. I previously used it because I thought whoever was using should be responsible for the access tokens, but now I think they shouldn't. The clients should manage their own tokens and if a request fails with a 400 status: they should be able to check that, update the token and retry the previously failed request. To simplify this we could have 2 HTTP clients one that just calls whatever upstream endpoint to request a new token and one that does all the usual stuff. 

Basically, this should be replaced with a switch statement 
* that checks if we have a 400, save our previous request,
* make an "I want a new token" request, save that token for the future
* retry our request that failed earlier with the new token.
* 
https://github.com/quarksgroup/payment-client/blob/8918501d655e5b4dd570e84065df3706c63921c5/payment/driver/fdi/fdi.go#L82

**Comment by iradukunda1** [comment]
Okay!
but for this retry staff,
I think it would be good if we implement it using `RoundTripper` instead of doing it using switch.

File: airtel/transport.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+const (
+	SchemeBearer = "Bearer"
+	SchemeToken  = "token"
+)
+
+//Transport  is an http.RoundTrip that refreshes oauth
+// tokens, wrapping a base Transport and refreshing the
+// token if expired with max retry.
+//request that require authorization header by appending header on it roundTripper
+type Transport struct {
```
Comment (rugwirobaker) [inline airtel/transport.go:22]: This is a retry transport right? Be specific about it then

Comment (iradukunda1) [reply airtel/transport.go:22]: okay! sure i will explain on it in the next changes

Comment (rugwirobaker) [reply airtel/transport.go:22]: I meant the name should be something like reTry transport, also there some good retry backoff libraries out there you can use: https://github.com/superfly/flyctl/blob/0247e3ef96acc66949a397a0e74820633cf4d2a5/retry/retry.go#L9-L30

File: airtel/driver/airtel/account_test.go [hunk 2]
```diff
@@ -21,7 +21,7 @@ func TestBalance(t *testing.T) {
 		Reply(200).
 		Type("application/json").
 		File("testdata/account.json")
-	client := NewDefault("encrypted-pin")
+	client := NewDefault("encrypted-pin", "client_id", "sceret", "grant_type")
```
Comment (rugwirobaker) [inline airtel/driver/airtel/account_test.go:24]: Small type `sceret` instead of `secret`

File: fdi/driver/fdi/fdi.go [hunk 3]
```diff
@@ -34,7 +34,7 @@ func New(uri, callback string, retry int) (*fdi.Client, error) {
 		return nil, err
 	}
 
-	transport := &fdi.Transport{
+	transport := &fdi.ReteryTransport{
```
Comment (rugwirobaker) [inline fdi/driver/fdi/fdi.go:37]: Typo: `ReteryTransport` instead of `RetryTransport`

File: airtel/driver/driver.go [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package driver
```
Comment (rugwirobaker) [inline airtel/driver/driver.go:1]: Do we need this driver package for anything? Just remove it completely.

Comment (iradukunda1) [reply airtel/driver/driver.go:1]: okay
